Mathlib/Topology/UniformSpace/Separated.lean

```diff
@@ -0,0 +1,60 @@
+/-
+Copyright (c) 2025 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Data.Set.Pairwise.Basic
+import Mathlib.Topology.UniformSpace.Defs
+
+/-!
+# Uniform separation
+
+This file defines a notion of separation of a set relative to an entourage.
+
+For an entourage `U`, a `U`-separated set `s` is a set such that every pair of elements of `s` is
+`U`-far.
+
+The concept of uniformly separated sets is used to define two further notions of separation:
+* Metric separation: `Metric.IsSeparated`, defined using the distance entourage.
+* Dynamical nets: `Dynamics.IsDynNetIn`, defined using the dynamical entourage.
+
+## TODO
+
+* Actually use `UniformSpace.IsSeparated` to define the above two notions.
+* Link to the notion of separation given by pairwise disjoint balls.
+-/
+
+open Set
+
+namespace UniformSpace
+variable {X : Type*} {U V : Set (X × X)} {s t : Set X} {x : X}
+
+/-- Given a uniformity `U`, a set `s` is `U`-separated if its elements are pairwise `U`-far from
+each other. -/
+def IsSeparated (U : Set (X × X)) (s : Set X) : Prop := s.Pairwise fun x y ↦ (x, y) ∉ U
+
+protected lemma IsSeparated.empty : IsSeparated U (∅ : Set X) := pairwise_empty _
+protected lemma IsSeparated.singleton : IsSeparated U {x} := pairwise_singleton ..
+
+@[simp] lemma IsSeparated.of_subsingleton (hs : s.Subsingleton) : IsSeparated U s := hs.pairwise _
+
+alias _root_.Set.Subsingleton.uniformSpaceIsSeparated := IsSeparated.of_subsingleton
+
+nonrec lemma IsSeparated.anti_left (hUV : U ⊆ V) (hs : IsSeparated V s) : IsSeparated U s :=
+  hs.mono' fun _x _y hxy h ↦ hxy <| hUV h
+
+lemma IsSeparated.anti_right (hst : s ⊆ t) (hs : IsSeparated U t) : IsSeparated U s := hs.mono hst
+
+lemma isSeparated_insert (hU : IsSymmetricRel U) :
+    IsSeparated U (insert x s) ↔ IsSeparated U s ∧ ∀ y ∈ s, (x, y) ∈ U → x = y := by
+  simpa [not_imp_not, IsSeparated] using pairwise_insert_of_symmetric fun _ _ ↦ hU.mk_mem_comm.not.1
```

Comment:
I've seen the discussion with Bhavik, but I still think we should have the versions without symmetry. In fact, your definition is already symmetric (in the sense that you can replace U by its symmetric), and stable under *outer* symmetrization (union with the symmetric), which is not "bad quantitatively"